File: services/PlexClient.ts

import { MediaClient } from './MediaClient';
import { EmbyItem, EmbyLibrary, FeedType, ServerConfig, VideoResponse, OrientationMode, SubtitleTrack, RecommendationCategory, WatchedHistoryItem, PlayQueue, PlayQueueItem, PlayMode } from '../types';

export class PlexClient extends MediaClient {
    
    private getCleanUrl() {
        return this.config.url.replace(/\/$/, "");
    }

    private getHeaders() {
        return {
            'Accept': 'application/json',
            'X-Plex-Token': this.config.token
        };
    }

    private async getMachineIdentifier(): Promise<string> {
        if (this.config.userId && this.config.userId !== '1') return this.config.userId;
        try {
            const response = await fetch(`${this.getCleanUrl()}/identity`, { headers: this.getHeaders() });
            if (response.ok) {
                const data = await response.json();
                return data.MediaContainer.machineIdentifier || data.MediaContainer.MachineIdentifier || '1';
            }
        } catch (e) {
            // Ignore fetch errors
        }
        return '1';
    }

    async authenticate(username: string, password: string): Promise<ServerConfig> {
        const token = password; 
        const response = await fetch(`${this.getCleanUrl()}/identity`, {
            headers: { 'Accept': 'application/json', 'X-Plex-Token': token }
        });
        if (!response.ok) throw new Error('Plex Connection Failed');
        const data = await response.json();
        const machineIdentifier = data.MediaContainer.machineIdentifier || data.MediaContainer.MachineIdentifier;
        return { url: this.config.url, username: username || 'Plex User', userId: machineIdentifier || '1', token: token, serverType: 'plex' };
    }

    async getLibraries(): Promise<EmbyLibrary[]> {
        const response = await fetch(`${this.getCleanUrl()}/library/sections`, { headers: this.getHeaders() });
        const data = await response.json();
        return data.MediaContainer.Directory.map((d: any) => ({ 
            Id: d.key, 
            Name: d.title, 
            CollectionType: d.type 
        }));
    }

    async getResumeItems(): Promise<EmbyItem[]> {
        try {
            const response = await fetch(`${this.getCleanUrl()}/library/onDeck`, { headers: this.getHeaders() });
            const data = await response.json();
            return this.mapPlexItems(data.MediaContainer.Metadata || []).slice(0, 12);
        } catch (e) {
            return [];
        }
    }

    private filterItems(items: EmbyItem[], mode: OrientationMode): EmbyItem[] {
        if (mode === 'both') return items;
        return items.filter(item => {
            const type = (item.Type || '').toLowerCase();
            const isNavFolder = ['show', 'season', 'folder'].includes(type);
            if (isNavFolder) return true;

            const w = item.Width || 0;
            const h = item.Height || 0;
            if (w === 0) return true; 
            if (mode === 'vertical') return h >= w * 0.8; 
            return w > h;
        });
    }

    async getVideos(
        navParentId: string | undefined, 
        library: EmbyLibrary | null, 
        feedType: FeedType, 
        skip: number, 
        limit: number, 
        orientationMode: OrientationMode,
        includeIds?: string
    ): Promise<VideoResponse> {
        const libraryName = library ? library.Name : "收藏";
        
        if (feedType === 'favorites') {
            const playlist = await this.findPlaylist(libraryName);
            if (!playlist) return { items: [], nextStartIndex: 0, totalCount: 0 };
            const response = await fetch(`${this.getCleanUrl()}/playlists/${playlist.ratingKey}/items?X-Plex-Container-Start=0&X-Plex-Container-Size=2000`, { headers: this.getHeaders() });
            const data = await response.json();
            const mappedItems = this.mapPlexItems(data.MediaContainer.Metadata || []);
            const filtered = this.filterItems(mappedItems, orientationMode);
            const reversed = filtered.reverse();
            return { items: reversed.slice(skip, skip + limit), nextStartIndex: skip + limit, totalCount: reversed.length };
        }

        let url: string;
        const isLibraryRoot = navParentId === undefined || (library && navParentId === library.Id);

        if (!isLibraryRoot && navParentId) {
            url = `${this.getCleanUrl()}/library/metadata/${navParentId}/children?X-Plex-Container-Start=${skip}&X-Plex-Container-Size=${limit}`;
        } else {
            const sectionId = navParentId || library?.Id;
            if (!sectionId && includeIds) {
                url = `${this.getCleanUrl()}/library/sections/${includeIds.split(',')[0]}/all?X-Plex-Container-Start=${skip}&X-Plex-Container-Size=${limit}`;
            } else if (!sectionId) {
                return { items: [], nextStartIndex: 0, totalCount: 0 };
            } else {
                let sort = 'addedAt:desc';
                if (feedType === 'random') sort = 'random';
                url = `${this.getCleanUrl()}/library/sections/${sectionId}/all?sort=${sort}&X-Plex-Container-Start=${skip}&X-Plex-Container-Size=${limit}`;
            }
        }

        const response = await fetch(url, { headers: this.getHeaders() });
        const data = await response.json();
        const items = data.MediaContainer.Metadata || [];
        const mappedItems = this.mapPlexItems(items);
        const filtered = this.filterItems(mappedItems, orientationMode);
        
        return { 
            items: filtered, 
            nextStartIndex: skip + items.length, 
            totalCount: data.MediaContainer.totalSize || data.MediaContainer.size || 0 
        };
    }

    private mapPlexItems(items: any[]): EmbyItem[] {
        return items.map((p: any) => {
             const media = p.Media?.[0];
             let formattedName = p.title;
             
             if (p.type === 'episode') {
                 const idx = p.index !== undefined ? String(p.index).padStart(2, '0') : '--';
                 formattedName = `${idx}. ${p.title}`;
             }

             return { 
                Id: p.ratingKey, 
                Name: formattedName, 
                Type: p.type, 
                MediaType: 'Video', 
                Overview: p.summary, 
                ProductionYear: p.year, 
                Width: media?.width, 
                Height: media?.height, 
                RunTimeTicks: p.duration ? p.duration * 10000 : undefined, 
                ImageTags: { Primary: p.thumb ? 'true' : undefined }, 
                UserData: {
                    IsFavorite: false,
                    PlayCount: p.viewCount || 0,
                    Played: (p.viewCount || 0) > 0,
                    PlaybackPositionTicks: p.viewOffset ? p.viewOffset * 10000 : 0,
                    LastPlayedDate: p.lastViewedAt ? new Date(p.lastViewedAt * 1000).toISOString() : undefined
                },
                _PlexThumb: p.thumb, 
                _PlexKey: media?.Part?.[0]?.key 
             };
        });
    }

    getVideoUrl(item: EmbyItem): string {
        const plexItem = item as any;
        if (plexItem._PlexKey) return `${this.getCleanUrl()}${plexItem._PlexKey}?X-Plex-Token=${this.config.token}`;
        return `${this.getCleanUrl()}/video/:/transcode/universal/start?path=${encodeURIComponent('/library/metadata/' + item.Id)}&mediaIndex=0&partIndex=0&protocol=hls&offset=0&fastSeek=1&directPlay=0&directStream=1&subtitleSize=100&audioBoost=100&X-Plex-Token=${this.config.token}`;
    }

    getImageUrl(itemId: string, tag?: string, type?: 'Primary' | 'Backdrop'): string {
        return `${this.getCleanUrl()}/photo/:/transcode?url=${encodeURIComponent(`/library/metadata/${itemId}/thumb`)}&width=800&height=1200&X-Plex-Token=${this.config.token}`;
    }

    private async findPlaylist(libraryName: string): Promise<any | null> {
        const title = `Tok-${libraryName}`;
        try {
            const response = await fetch(`${this.getCleanUrl()}/playlists?title=${encodeURIComponent(title)}`, { headers: this.getHeaders() });
            const data = await response.json();
            return data.MediaContainer.Metadata?.find((p: any) => p.title === title) || null;
        } catch (e) { return null; }
    }

    async getFavorites(libraryName: string): Promise<Set<string>> {
        const playlist = await this.findPlaylist(libraryName);
        if (!playlist) return new Set();
        try {
            const response = await fetch(`${this.getCleanUrl()}/playlists/${playlist.ratingKey}/items?X-Plex-Container-Size=2000`, { headers: this.getHeaders() });
            const data = await response.json();
            return new Set((data.MediaContainer.Metadata || []).map((i: any) => i.ratingKey));
        } catch (e) { return new Set(); }
    }

    async toggleFavorite(itemId: string, isFavorite: boolean, libraryName: string): Promise<void> {
        const playlist = await this.findPlaylist(libraryName);
        const machineId = await this.getMachineIdentifier();
        const itemUri = `server://${machineId}/com.plexapp.plugins.library/library/metadata/${itemId}`;
        if (isFavorite) {
            if (!playlist) return;
            const itemsRes = await fetch(`${this.getCleanUrl()}/playlists/${playlist.ratingKey}/items?X-Plex-Container-Size=2000`, { headers: this.getHeaders() });
            const entry = (await itemsRes.json()).MediaContainer.Metadata?.find((i: any) => i.ratingKey === itemId);
            if (entry?.playlistItemID) {
                await fetch(`${this.getCleanUrl()}/playlists/${playlist.ratingKey}/items/${entry.playlistItemID}?X-Plex-Token=${this.config.token}`, { method: 'DELETE', headers: this.getHeaders() });
            }
        } else {
            if (playlist) {
                await fetch(`${this.getCleanUrl()}/playlists/${playlist.ratingKey}/items?uri=${encodeURIComponent(itemUri)}&X-Plex-Token=${this.config.token}`, { method: 'PUT', headers: this.getHeaders() });
            } else {
                await fetch(`${this.getCleanUrl()}/playlists?type=video&title=${encodeURIComponent(`Tok-${libraryName}`)}&smart=0&uri=${encodeURIComponent(itemUri)}&X-Plex-Token=${this.config.token}`, { method: 'POST', headers: this.getHeaders() });
            }
        }
    }

    async deleteItem(itemId: string): Promise<void> {
        throw new Error('Plex does not support delete functionality');
    }

    async searchItems(query: string): Promise<EmbyItem[]> {
        try {
            const response = await fetch(`${this.getCleanUrl()}/search?query=${encodeURIComponent(query)}`, { 
                headers: this.getHeaders() 
            });
            
            const data = await response.json();
            const items = data.MediaContainer?.Metadata || [];
            const videoItems = items.filter((item: any) => 
                ['movie', 'episode', 'show', 'video'].includes(item.type)
            );
            
            return this.mapPlexItems(videoItems);
        } catch (error) {
            console.error('Plex search failed:', error);
            return [];
        }
    }

    async getSubtitleTracks(itemId: string): Promise<SubtitleTrack[]> {
        try {
            const response = await fetch(`${this.getCleanUrl()}/library/metadata/${itemId}`, { 
                headers: this.getHeaders() 
            });
            
            const data = await response.json();
            const metadata = data.MediaContainer?.Metadata?.[0];
            const subtitleTracks: SubtitleTrack[] = [];
            
            if (metadata?.Media) {
                metadata.Media.forEach((media: any) => {
                    if (media.Part) {
                        media.Part.forEach((part: any) => {
                            if (part.Stream) {
                                part.Stream
                                    .filter((stream: any) => stream.streamType === 3) // 3 = subtitle
                                    .forEach((stream: any) => {
                                        let url: string | undefined;
                                        if (stream.key) {
                                            url = `${this.getCleanUrl()}${stream.key}?X-Plex-Token=${this.config.token}`;
                                        }
                                        
                                        subtitleTracks.push({
                                            id: `${itemId}_${stream.id}`,
                                            label: stream.displayTitle || stream.languageCode || `字幕 ${subtitleTracks.length + 1}`,
                                            language: stream.languageCode || 'und',
                                            isDefault: stream.default || false,
                                            codec: stream.codec,
                                            isExternal: false,
                                            url
                                        });
                                    });
                            }
                        });
                    }
                });
            }
            
            return subtitleTracks;
        } catch (error) {
            console.error('Failed to get subtitle tracks:', error);
            return [];
        }
    }

    /**
     * Plex 暂不支持 Recommendations API
     * 返回空数组
     */
    async getRecommendations(): Promise<RecommendationCategory[]> {
        // Plex API 不提供与 Emby 相同的推荐功能
        return [];
    }

    /**
     * Plex 暂不支持播放进度同步到服务器
     * @param itemId 媒体项ID
     * @param positionTicks 播放位置
     */
    async updatePlaybackProgress(itemId: string, positionTicks: number): Promise<void> {
        // Plex 不支持此功能，保持本地状态
        console.warn('Plex 不支持远程播放进度同步');
    }

    /**
     * Plex 暂不支持用户评分
     * @param itemId 媒体项ID
     * @param rating 评分（0-10）
     */
    async updateUserRating(itemId: string, rating: number): Promise<void> {
        // Plex 不支持此功能
        console.warn('Plex 不支持用户评分同步');
    }

    /**
     * Plex 获取观影历史
     * 通过已观看的媒体获取历史记录
     * @param limit 返回记录数量限制
     */
    async getWatchedHistory(limit: number = 50): Promise<WatchedHistoryItem[]> {
        try {
            // Plex 使用 /library/allLeaves?viewedAt=1 获取已观看项目
            const response = await fetch(
                `${this.getCleanUrl()}/library/all?type=4&viewedAt=1&X-Plex-Container-Start=0&X-Plex-Container-Size=${limit}`,
                { headers: this.getHeaders() }
            );
            
            const data = await response.json();
            const items = data.MediaContainer?.Metadata || [];
            
            return items.map((item: any) => ({
                id: item.ratingKey,
                itemId: item.ratingKey,
                name: item.title || '未命名',
                type: item.type,
                mediaType: 'Video',
                overview: item.summary,
                productionYear: item.year,
                width: item.Media?.[0]?.width,
                height: item.Media?.[0]?.height,
                runTimeTicks: item.duration ? item.duration * 10000 : 0,
                playbackPositionTicks: item.viewOffset ? item.viewOffset * 10000 : 0,
                playCount: item.viewCount || 0,
                played: (item.viewCount || 0) > 0,
                lastPlayedDate: item.lastViewedAt ? new Date(item.lastViewedAt * 1000).toISOString() : undefined,
                isFavorite: false
            }));
        } catch (error) {
            console.error('获取观影历史失败:', error);
            return [];
        }
    }

    /**
     * Plex 标记为已观看（不支持远程同步）
     * @param itemId 媒体项ID
     */
    async markAsWatched(itemId: string): Promise<void> {
        // Plex 不支持远程标记已观看
        console.warn('Plex 不支持远程标记已观看');
    }

    /**
     * Plex 标记为未观看（不支持远程同步）
     * @param itemId 媒体项ID
     */
    async markAsUnwatched(itemId: string): Promise<void> {
        // Plex 不支持远程标记未观看
        console.warn('Plex 不支持远程标记未观看');
    }

    // ==================== 播放队列相关实现 ====================

    /**
     * 创建播放队列
     * Plex 不支持播放队列 API，使用本地队列模拟
     */
    async createPlayQueue(items: EmbyItem[], startIndex: number = 0): Promise<PlayQueue> {
        const now = Date.now();
        const queueId = `local_queue_${now}`;

        const queueItems: PlayQueueItem[] = items.map((item, index) => ({
            id: `queue_item_${now}_${index}`,
            item: item,
            addedAt: now
        }));

        return {
            id: queueId,
            name: 'EmbyTok Queue',
            items: queueItems,
            currentIndex: startIndex,
            playMode: PlayMode.Sequential,
            createdAt: now,
            updatedAt: now
        };
    }

    /**
     * 添加项目到播放队列
     */
    async addToPlayQueue(queueId: string, items: EmbyItem[]): Promise<void> {
        console.log(`[PlexClient] 添加 ${items.length} 个项目到队列 ${queueId}`);
    }

    /**
     * 从播放队列移除项目
     */
    async removeFromPlayQueue(queueId: string, itemIds: string[]): Promise<void> {
        console.log(`[PlexClient] 从队列 ${queueId} 移除项目: ${itemIds.join(', ')}`);
    }

    /**
     * 获取剧集的下一集
     * 调用 Plex 的 /shows/{ratingKey}/allLeaves 获取剧集列表
     */
    async getNextEpisode(seriesId: string, currentEpisodeId: string): Promise<EmbyItem | null> {
        try {
            const response = await fetch(
                `${this.getCleanUrl()}/shows/${seriesId}/allLeaves?X-Plex-Container-Start=0&X-Plex-Container-Size=500`,
                { headers: this.getHeaders() }
            );

            if (!response.ok) {
                console.error(`获取剧集列表失败: ${response.status}`);
                return null;
            }

            const data = await response.json();
            const episodes = data.MediaContainer?.Metadata || [];

            if (episodes.length === 0) {
                return null;
            }

            // 找到当前剧集的索引
            const currentIndex = episodes.findIndex((ep: any) => ep.ratingKey === currentEpisodeId);

            if (currentIndex === -1) {
                console.warn(`未找到当前剧集 ${currentEpisodeId}`);
                return null;
            }

            // 获取下一集
            const nextIndex = currentIndex + 1;

            if (nextIndex >= episodes.length) {
                return null;
            }

            const nextEp = episodes[nextIndex];
            const mapped = this.mapPlexItems([nextEp]);

            return mapped[0] || null;
        } catch (error) {
            console.error('获取下一集失败:', error);
            return null;
        }
    }

    /**
     * 获取播放队列
     * Plex 不支持播放队列，返回空队列
     */
    async getPlayQueue(queueId: string): Promise<PlayQueue> {
        return {
            id: queueId,
            name: 'EmbyTok Queue',
            items: [],
            currentIndex: 0,
            playMode: PlayMode.Sequential,
            createdAt: Date.now(),
            updatedAt: Date.now()
        };
    }
}

File: android-kmp/shared/network/src/commonMain/kotlin/com/embytok/network/plex/PlexClient.kt
package com.embytok.network.plex

import com.embytok.common.EmbyTokLogger
import com.embytok.domain.client.ImageType
import com.embytok.domain.client.MediaClient
import com.embytok.domain.client.PlayMode
import com.embytok.domain.model.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Plex 服务器客户端实现
 */
class PlexClient(
    override val config: ServerConfig,
    private val httpClient: HttpClient
) : MediaClient {

    private val logger = EmbyTokLogger
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val baseUrl: String
        get() = config.getFullUrl()

    private val authHeader: HeadersBuilder.() -> Unit = {
        append("X-Plex-Token", config.token)
    }

    override suspend fun authenticate(username: String, password: String): ServerConfig {
        logger.d("Plex 认证（Token 模式）", "PlexClient")

        // Plex 使用 X-Plex-Token 直接认证，这里验证连接
        val identity = httpClient.get("$baseUrl/identity") {
            authHeader()
        }.body<PlexIdentityResponse>()

        if (identity.MediaContainer?.machineIdentifier == null) {
            throw IllegalStateException("Invalid Plex server")
        }

        return config.copy(
            serverName = identity.MediaContainer.friendlyName ?: "Plex"
        )
    }

    override suspend fun getLibraries(): List<EmbyLibrary> {
        logger.d("获取 Plex 媒体库", "PlexClient")

        val response = httpClient.get("$baseUrl/library/sections") {
            authHeader()
        }.body<PlexDirectoryResponse>()

        return response.MediaContainer?.Directory?.map { dir ->
            EmbyLibrary(
                Id = dir.key ?: "",
                Name = dir.title ?: "",
                collectionType = when (dir.type) {
                    "movie" -> "movies"
                    "show" -> "tvshows"
                    "artist" -> "music"
                    else -> dir.type
                }
            )
        } ?: emptyList()
    }

    override suspend fun getResumeItems(): List<EmbyItem> {
        logger.d("获取 Plex 继续观看", "PlexClient")

        // Plex 的"继续观看"在 /hubs 中
        val response = httpClient.get("$baseUrl/hubs") {
            authHeader()
            parameter("continue", true)
        }.body<PlexHubsResponse>()

        return response.MediaContainer?.Hub?.flatMap { hub ->
            hub.Metadata?.mapNotNull { meta -> meta.toEmbyItem() } ?: emptyList()
        } ?: emptyList()
    }

    override suspend fun getVideos(
        parentId: String?,
        library: EmbyLibrary?,
        feedType: FeedType,
        skip: Int,
        limit: Int,
        orientationMode: OrientationMode,
        includeIds: List<String>?
    ): VideoResponse {
        logger.d("获取 Plex 视频: parentId=$parentId, feedType=$feedType", "PlexClient")

        val sectionKey = parentId ?: library?.Id ?: return VideoResponse(emptyList(), 0)

        val sortParam = when (feedType) {
            FeedType.LATEST -> "addedAt:desc"
            FeedType.RANDOM -> "random"
            else -> "titleSort"
        }

        val response = httpClient.get("$baseUrl/library/sections/$sectionKey/all") {
            authHeader()
            parameter("sort", sortParam)
            parameter("limit", limit)
            parameter("offset", skip)
        }.body<PlexMetadataResponse>()

        val items = response.MediaContainer?.Metadata?.mapNotNull { it.toEmbyItem() } ?: emptyList()
        val total = response.MediaContainer?.totalSize?.toIntOrNull() ?: items.size

        val filtered = applyOrientationFilter(items, orientationMode)

        return VideoResponse(filtered, total)
    }

    override suspend fun getVideoUrl(item: EmbyItem, mode: PlayMode): String {
        logger.d("获取 Plex 视频 URL: item=${item.Name}, mode=$mode", "PlexClient")

        return when (mode) {
            PlayMode.DIRECT -> {
                // Plex 直链使用 _PlexKey
                item._PlexKey ?: throw IllegalStateException("No Plex key available")
            }
            PlayMode.TRANSCODE, PlayMode.FALLBACK -> {
                // Plex 转码
                val bitrate = if (mode == PlayMode.FALLBACK) 2000 else 0
                val params = buildString {
                    append("path=")
                    append(item._PlexKey?.encodeURLParameter() ?: "")
                    append("&mediaBitrate=")
                    append(bitrate)
                    append("&protocol=web")
                }
                "$baseUrl/video/:/transcode/universal/start?$params"
            }
        }
    }

    override suspend fun getImageUrl(itemId: String, tag: String?, type: ImageType): String {
        // Plex 图片使用特殊的 URL 格式
        val ratingKey = tag ?: itemId
        return when (type) {
            ImageType.PRIMARY -> "$baseUrl/library/metadata/$ratingKey/thumb?width=800&height=1200"
            ImageType.BACKDROP -> "$baseUrl/library/metadata/$ratingKey/art?width=1920&height=1080"
            ImageType.THUMB -> "$baseUrl/library/metadata/$ratingKey/thumb?width=400&height=300"
            ImageType.LOGO -> "$baseUrl/library/metadata/$ratingKey/art?width=800"
        } + "&X-Plex-Token=${config.token}"
    }

    override suspend fun getFavorites(libraryName: String): Set<String> {
        logger.d("获取 Plex 收藏: library=$libraryName", "PlexClient")

        // Plex 收藏使用播放列表
        val response = httpClient.get("$baseUrl/playlists") {
            authHeader()
            parameter("playlistType", "video")
        }.body<PlexPlaylistResponse>()

        val playlist = response.MediaContainer?.Playlist?.find {
            it.title == "Tok-$libraryName"
        } ?: return emptySet()

        // 获取播放列表中的项
        val items = httpClient.get("$baseUrl/playlists/${playlist.key}/items") {
            authHeader()
        }.body<PlexMetadataResponse>()

        return items.MediaContainer?.Metadata?.mapNotNull { it.ratingKey }?.toSet() ?: emptySet()
    }

    override suspend fun toggleFavorite(itemId: String, isFavorite: Boolean, libraryName: String) {
        logger.d("切换 Plex 收藏: itemId=$itemId, isFavorite=$isFavorite", "PlexClient")

        val action = if (isFavorite) "favorite" else "unfavorite"
        httpClient.get("$baseUrl/library/metadata/$itemId/$action") {
            authHeader()
        }
    }

    override suspend fun deleteItem(itemId: String) {
        logger.d("删除 Plex 媒体项: $itemId", "PlexClient")

        httpClient.delete("$baseUrl/library/metadata/$itemId") {
            authHeader()
        }
    }

    override suspend fun searchItems(query: String): List<EmbyItem> {
        logger.d("Plex 搜索: $query", "PlexClient")

        val response = httpClient.get("$baseUrl/search") {
            authHeader()
            parameter("query", query)
        }.body<PlexSearchResponse>()

        return response.MediaContainer?.Metadata?.mapNotNull { it.toEmbyItem() } ?: emptyList()
    }

    override suspend fun getSubtitleTracks(itemId: String): List<SubtitleTrack> {
        logger.d("获取 Plex 字幕轨道: $itemId", "PlexClient")

        val response = httpClient.get("$baseUrl/library/metadata/$itemId") {
            authHeader()
        }.body<PlexMetadataResponse>()

        val media = response.MediaContainer?.Metadata?.firstOrNull()?.Media?.firstOrNull()
        val streams = media?.Part?.flatMap { part ->
            part.Stream?.filter { it.streamType == 3 }?.map { stream ->
                SubtitleTrack(
                    id = stream.id?.toString() ?: "0",
                    label = stream.displayTitle ?: stream.language ?: "Subtitle",
                    srclang = stream.language,
                    src = "${baseUrl}${stream.key}?X-Plex-Token=${config.token}",
                    type = when (stream.codec?.lowercase()) {
                        "srt" -> "srt"
                        "ssa", "ass" -> "ass"
                        else -> "vtt"
                    }
                )
            } ?: emptyList()
        } ?: emptyList()

        return streams
    }

    private fun applyOrientationFilter(items: List<EmbyItem>, mode: OrientationMode): List<EmbyItem> {
        if (mode == OrientationMode.BOTH) return items

        return items.filter { item ->
            when (mode) {
                OrientationMode.VERTICAL -> item.isVertical()
                OrientationMode.HORIZONTAL -> item.isHorizontal()
                else -> true
            }
        }
    }

    // ============ Plex 响应模型 ============

    @Serializable
    private data class PlexIdentityResponse(
        val MediaContainer: PlexIdentity?
    )

    @Serializable
    private data class PlexIdentity(
        val machineIdentifier: String? = null,
        val friendlyName: String? = null
    )

    @Serializable
    private data class PlexDirectoryResponse(
        val MediaContainer: PlexDirectoryContainer?
    )

    @Serializable
    private data class PlexDirectoryContainer(
        val Directory: List<PlexDirectory> = emptyList()
    )

    @Serializable
    private data class PlexDirectory(
        val key: String? = null,
        val title: String? = null,
        val type: String? = null
    )

    @Serializable
    private data class PlexHubsResponse(
        val MediaContainer: PlexHubsContainer?
    )

    @Serializable
    private data class PlexHubsContainer(
        val Hub: List<PlexHub> = emptyList()
    )

    @Serializable
    private data class PlexHub(
        val Metadata: List<PlexMetadata>? = null
    )

    @Serializable
    private data class PlexMetadataResponse(
        val MediaContainer: PlexMetadataContainer?
    )

    @Serializable
    private data class PlexMetadataContainer(
        val Metadata: List<PlexMetadata> = emptyList(),
        val totalSize: String? = null
    )

    @Serializable
    private data class PlexMetadata(
        val ratingKey: String? = null,
        val key: String? = null,
        val title: String? = null,
        val type: String? = null,
        val summary: String? = null,
        val year: Int? = null,
        val duration: Long? = null,
        val thumb: String? = null,
        val art: String? = null,
        val Media: List<PlexMedia>? = null
    ) {
        fun toEmbyItem(): EmbyItem? {
            if (ratingKey == null || title == null) return null

            val media = Media?.firstOrNull()
            val videoStream = media?.Part?.firstOrNull()?.Stream?.find { it.streamType == 1 }
            val width = videoStream?.width?.toIntOrNull()
            val height = videoStream?.height?.toIntOrNull()
            val duration = duration ?: media?.duration?.toLongOrNull()

            return EmbyItem(
                Id = ratingKey,
                Name = title,
                Type = when (type) {
                    "movie" -> "Movie"
                    "episode" -> "Episode"
                    "show" -> "Series"
                    "season" -> "Season"
                    "track" -> "Audio"
                    else -> "Video"
                },
                Overview = summary,
                ProductionYear = year,
                Width = width,
                Height = height,
                runTimeTicks = (duration ?: 0) * 10_000, // Plex 使用毫秒
                _PlexKey = key,
                ImageTags = ImageTags(
                    Primary = thumb?.let { "thumb" },
                    Backdrop = art?.let { "art" }
                )
            )
        }
    }

    @Serializable
    private data class PlexMedia(
        val duration: String? = null,
        val Part: List<PlexPart>? = null
    )

    @Serializable
    private data class PlexPart(
        val key: String? = null,
        val duration: String? = null,
        val Stream: List<PlexStream>? = null
    )

    @Serializable
    private data class PlexStream(
        val id: Int? = null,
        val streamType: Int? = null,
        val codec: String? = null,
        val language: String? = null,
        val displayTitle: String? = null,
        val width: String? = null,
        val height: String? = null,
        val key: String? = null
    )

    @Serializable
    private data class PlexPlaylistResponse(
        val MediaContainer: PlexPlaylistContainer?
    )

    @Serializable
    private data class PlexPlaylistContainer(
        val Playlist: List<PlexPlaylist> = emptyList()
    )

    @Serializable
    private data class PlexPlaylist(
        val ratingKey: String? = null,
        val key: String? = null,
        val title: String? = null
    )

    @Serializable
    private data class PlexSearchResponse(
        val MediaContainer: PlexSearchContainer?
    )

    @Serializable
    private data class PlexSearchContainer(
        val Metadata: List<PlexMetadata> = emptyList()
    )
}
